lib/active_merchant/billing/gateways/vantiv_express.rb

```diff
@@ -0,0 +1,475 @@
+require 'nokogiri'
+require 'securerandom'
+
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class VantivExpressGateway < Gateway
+      self.test_url = 'https://certtransaction.elementexpress.com'
+      self.live_url = 'https://transaction.elementexpress.com'
+
+      self.supported_countries = ['US']
+      self.default_currency = 'USD'
+      self.supported_cardtypes = %i[visa master american_express discover diners_club jcb]
+
+      self.homepage_url = 'http://www.elementps.com'
+      self.display_name = 'Element'
+
+      SERVICE_TEST_URL = 'https://certservices.elementexpress.com'
+      SERVICE_LIVE_URL = 'https://services.elementexpress.com'
+
+      NETWORK_TOKEN_TYPE = {
+        apple_pay: 2,
+        google_pay: 1
+      }
+
+      CARD_PRESENT_CODE = {
+        'Unknown' => 1,
+        'Present' => 2,
+        'NotPresent' => 3
+      }
+
+      MARKET_CODE = {
+        'AutoRental' => 1,
+        'DirectMarketing' => 2,
+        'ECommerce' => 3,
+        'FoodRestaurant' => 4,
+        'HotelLodging' => 5,
+        'Petroleum' => 6,
+        'Retail' => 7,
+        'QSR' => 8,
+        'Grocery' => 9
+      }
+
+      PAYMENT_TYPE = {
+        'NotUsed' => 0,
+        'Recurring' => 1,
+        'Installment' => 2,
+        'CardHolderInitiated' => 3,
+        'CredentialOnFile' => 4
+      }
+
+      REVERSAL_TYPE = {
+        'System' => 0,
+        'Full' => 1,
+        'Partial' => 2
+      }
+
+      SUBMISSION_TYPE = {
+        'NotUsed' => 0,
+        'Initial' => 1,
+        'Subsequent' => 2,
+        'Resubmission' => 3,
+        'ReAuthorization' => 4,
+        'DelayedCharges' => 5,
+        'NoShow' => 6
+      }
+
+      def initialize(options = {})
+        requires!(options, :account_id, :account_token, :application_id, :acceptor_id, :application_name, :application_version)
+        super
+      end
+
+      def purchase(money, payment, options = {})
+        action = payment.is_a?(Check) ? 'CheckSale' : 'CreditCardSale'
+        eci = payment.is_a?(NetworkTokenizationCreditCard) ? parse_eci(payment) : nil
+
+        request = build_xml_request do |xml|
+          xml.send(action, xmlns: live_url) do
+            add_credentials(xml)
```

Comment:
Should we move `add_credentials` to `def commit` since it's used on all requests?